ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/DatabaseSelfServiceError.scala

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.appendonlydao
+
+import com.daml.error.ContextualizedErrorLogger
+import com.daml.error.definitions.IndexErrors
+import io.grpc.StatusRuntimeException
+import org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException
+import org.postgresql.util.PSQLException
+
+import java.sql._
+import scala.annotation.tailrec
+
+/** Wraps SQLExceptions into transient and non-transient errors.
+  * Transience classification is done as follows:
+  *
+  * * Problems that are likely to be resolved with retries are transient errors. For example, network outages or db access
+  * serialization problems.
+  *
+  * * Problems that cannot be recovered from are non-transient. For example, an illegal argument exception inside a
+  * database transaction or a unique constraint violation.
+  *
+  * TODO error codes: Move and handle the error specialization per storage backend
+  *                   in [[com.daml.platform.store.backend.StorageBackend]].
+  */
+object DatabaseSelfServiceError {
+  @tailrec
+  def apply(
+      exception: Throwable
+  )(implicit contextualizedErrorLogger: ContextualizedErrorLogger): Throwable = exception match {
+    // This frequently occurs when running with H2, because H2 does not properly implement the serializable
+    // isolation level. This causes unexpected constraint violation exceptions when running with H2 in the presence
+    // of contention. For now, we retry on these exceptions.
+    // See https://github.com/h2database/h2database/issues/2167
+    case ex: JdbcSQLIntegrityConstraintViolationException => retryable(ex)
+    case ex: SQLRecoverableException => retryable(ex)
+    case ex: SQLTransientException => retryable(ex)
+    case ex: SQLNonTransientException => nonRetryable(ex)
+    case ex: PSQLException => if (isRetryablePsqlException(ex)) retryable(ex) else nonRetryable(ex)
```

Comment:
Normally I would suggest to push that down under StorageBackend. It was created to hide the specifics of used databases. (for example there is no isRetryableH2Exception)
Also I understand this is best effort try to solve the problem, right? I mean even in this PR there are inconsistencies like SQLNonTransientExceptions are handled as non-retry here, but retry in the initialization. Which makes sense to me: precisely it could be decided only on a case by case fashion. (needless to say: which would lead to an enormous explosion...multiplied by supported DB backends)

No need to change as it is best effort

Reply:
Thanks. Specializing exception handling behavior per StorageBackend implementation is a good point. (Already in a TODO in this file) 👍